### PR TITLE
sql: implement the `DROP TYPE` command

### LINKED ISSUE #48363: sql: support `DROP TYPE`
Support the `DROP TYPE` command. This should properly interact with name draining as is done on table descriptors.

### PR DESCRIPTION
**Pull request by rohany**
This PR implements the `DROP TYPE` command without support for
`CASCADE`. It lays the foundation of the dependency management between
types and tables that use them. With this, we can drop types when a
database is dropped.

Fixes #48363.

Release note: None

### CODE REVIEW

**Comment by rohany** [comment]
This looks like it needs the import changes in https://github.com/cockroachdb/cockroach/pull/51149.

File: pkg/sql/drop_type.go [hunk 1]
```diff
@@ -12,27 +12,168 @@ package sql
...
-	)
+	for _, typ := range n.td {
+		if err := params.p.dropTypeImpl(params.ctx, typ, tree.AsStringWithFQNames(n.n, params.Ann())); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (p *planner) addTypeBackReference(
```
Comment (RichardJCai) [inline pkg/sql/drop_type.go:123]: I think a backref also has to be added for views.

```
root@127.0.0.1:65260/movr> create type test as enum ('hello');
CREATE TYPE

Time: 2.043ms

root@127.0.0.1:65260/movr> create view vx as SELECT 'hello'::test;
CREATE VIEW

Time: 3.804ms

root@127.0.0.1:65260/movr> drop type test;
DROP TYPE

Time: 27.613ms

root@127.0.0.1:65260/movr> create view vx as SELECT 'hello'::test;
root@127.0.0.1:65260/movr> select * from vx;
ERROR: type ID 59 in descriptor not found: descriptor not found
```

**Review by RichardJCai — CHANGES_REQUESTED** [review]

File: pkg/sql/drop_type.go (hunk 1 above)
Comment (rohany) [reply pkg/sql/drop_type.go:123]: Good catch, updated.

**Review by ajwerner — CHANGES_REQUESTED** [review]
Mostly lgtm just some commentary

> Reviewed 3 of 17 files at r1, 1 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51240#-:-MClXV3DBhl5mkIvJ5j4:bttd90i)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @lucy-zhang, @RichardJCai, and @rohany)

___
*[pkg/sql/schema_changer.go, line 470 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51240#-MClVocA6luu7JLN6fF6:-MClVocBAz-xSMk5TWeV:bpi9g0n) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8a848b4d4e2f871a7b80e08e35ffccc1f7b85cc/pkg/sql/schema_changer.go#L470)):*
> ```Go
> 		// by the table.
> 		// TODO (rohany): Do we need to delay the back reference removal until
> 		//  the schema change job execution? Doing it here leads means that we
> ```

`leads means` but more to the point, I think we need to sort this out. 

Why does doing it here mean that? Can't we be robust to the concurrent operations?
___
*[pkg/sql/schema_changer.go, line 1037 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51240#-MClWZV75qZT033GcxHs:-MClWZV75qZT033GcxHt:bz325gg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8a848b4d4e2f871a7b80e08e35ffccc1f7b85cc/pkg/sql/schema_changer.go#L1037)):*
> ```Go
> 		if !scTable.Dropped() {
> 			newReferencedTypeIDs, err := scTable.GetAllReferencedTypeIDs(func(id sqlbase.ID) (*TypeDescriptor, error) {
> 				return descs[id].(*sqlbase.MutableTypeDescriptor).TypeDesc(), nil
> ```

Are you worried about panics here? 
___
*[pkg/sql/type_change.go, line 161 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51240#-MClXOvZ-kX7JAW5S0X7:-MClXOvZ-kX7JAW5S0X8:brx1zpm) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8a848b4d4e2f871a7b80e08e35ffccc1f7b85cc/pkg/sql/type_change.go#L161)):*
> ```Go
> 	}
> 
> 	// TODO (rohany): Not sure if this needs to be done before or after
> ```

I think after draining. 
___
*[pkg/sql/type_change.go, line 166 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51240#-MClX6QiCW8h5vrDqkAH:-MClX6QiCW8h5vrDqkAI:bo2infx) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8a848b4d4e2f871a7b80e08e35ffccc1f7b85cc/pkg/sql/type_change.go#L166)):*
> ```Go
> 	if typeDesc.State == sqlbase.TypeDescriptor_DROP {
> 		if err := t.execCfg.DB.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
> 			// TODO (rohany): Do we need to set the systemConfigTrigger here?
> ```

Nope. We need to do that when creating a split point (table) or changing zone configs. Working on an RFC on that topic right now.

**Review by rohany — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51240)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @lucy-zhang, and @RichardJCai)

___
*[pkg/sql/schema_changer.go, line 470 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51240#-MClVocA6luu7JLN6fF6:-MCljhgcCi_wB0q1plWs:b51xutb) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8a848b4d4e2f871a7b80e08e35ffccc1f7b85cc/pkg/sql/schema_changer.go#L470)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

`leads means` but more to the point, I think we need to sort this out. 

Why does doing it here mean that? Can't we be robust to the concurrent operations?
</blockquote></details>

I meant doing the back referenced removal in the transaction that commits the drop table vs removing the back reference in the schema changer. If we delay to the schema changer then we can't do 

```
BEGIN;
DROP TABLE t;
DROP TYPE <t used to use>
END;
```
I'm not sure that there is a problem with doing it in the drop table txn.
___
*[pkg/sql/schema_changer.go, line 1037 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51240#-MClWZV75qZT033GcxHs:-MClkNxT9LRVOpyQ37nU:blvctbr) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8a848b4d4e2f871a7b80e08e35ffccc1f7b85cc/pkg/sql/schema_changer.go#L1037)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Are you worried about panics here? 
</blockquote></details>

No, because the descs map is a `MutableDescriptor` map
___
*[pkg/sql/type_change.go, line 161 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51240#-MClXOvZ-kX7JAW5S0X7:-MClkY-j7ll3Jh0tTY_T:b-zcjvte) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8a848b4d4e2f871a7b80e08e35ffccc1f7b85cc/pkg/sql/type_change.go#L161)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I think after draining. 
</blockquote></details>

Thanks
___
*[pkg/sql/type_change.go, line 166 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51240#-MClX6QiCW8h5vrDqkAH:-MClkXSc5ejw-W0ECsGp:b-zcjvte) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8a848b4d4e2f871a7b80e08e35ffccc1f7b85cc/pkg/sql/type_change.go#L166)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Nope. We need to do that when creating a split point (table) or changing zone configs. Working on an RFC on that topic right now. 
</blockquote></details>

Thanks

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51240)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @lucy-zhang, @RichardJCai, and @rohany)

___
*[pkg/sql/schema_changer.go, line 470 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51240#-MClVocA6luu7JLN6fF6:-MClsG-3AAQaQGPwRw1Q:bt8n8g0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8a848b4d4e2f871a7b80e08e35ffccc1f7b85cc/pkg/sql/schema_changer.go#L470)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I meant doing the back referenced removal in the transaction that commits the drop table vs removing the back reference in the schema changer. If we delay to the schema changer then we can't do 

```
BEGIN;
DROP TABLE t;
DROP TYPE <t used to use>
END;
```
I'm not sure that there is a problem with doing it in the drop table txn.
</blockquote></details>

I think I'm confused. Can't we just make sure that one of the two operations cleans up the back reference?

**Comment by rohany** [comment]
> I think I'm confused. Can't we just make sure that one of the two operations cleans up the back reference?

Yeah we can, I just wasn't sure in which place we should do it. We had a small discussion about moving it after the drop txn commits to maybe avoid that publish multiple stale state issue, but this wasn't the problem.

**Comment by rohany** [comment]
Updated to do the backreference removal in the `DROP TABLE` txn.

**Comment by rohany** [comment]
cc @knz for some comments you had:

pkg/sql/schema_changer.go, line 468 at r3 (raw file):

		// If this table is being dropped, remove all back references from
		// types that this table uses. First collect all types referenced
		// by the table.
I don't understand why this is happening only here. Why are the back references not removed as soon as the DROP statement executes, like we already drop the back references for views etc?

Lucy should help here.

-- This has been updated here.

pkg/sql/schema_changer.go, line 1033 at r3 (raw file):

		// Now that all mutations have been applied, find the new set of referenced
		// type descriptors. If this table has been dropped in the mean time, then
ditto

-- This case is a bit different. When dropping a particular schema element, we can only remove the back reference once the state machine through the schema changer finishes. Imagine dropping a column with a user defined type -- the column is still present on the table while it is in the `DELETE_ONLY` and `DELETE_AND_WRITE_ONLY` states.

pkg/sql/sqlbase/structured.go, line 1358 at r3 (raw file):

	}
	// Sort the output so that the order is deterministic.
	sort.Sort(result)

What's going on here? If you think this is necessary, it sounds to me as if it would have been necessary earlier (i.e. it's a bug fix for an outstanding issue). In that case I would request that you extract this into a separate commit, add the tests and refer to the issue that it closes.

-- This sorting is only became necessary here so that we add back references in the same order every time. This just ensures that the output list of dependent tables is always the same.

pkg/sql/sqlbase/type_desc.go, line 319 at r3 (raw file):

// AddReferencingDescriptorID adds a new referencing descriptor ID to the
// TypeDescriptor. It ensures that duplicates are not added.
This machinery should be shared for all inter-object references, not just UDTs.

-- I agree, but it's somewhat more difficult to do for tables than for types. For tables the backreference has to be inserted into the correct child object depending on what kind of reference is being made. For types we don't do such a thing.

**Comment by rohany** [comment]
Friendly ping!

**Review by thoszhang — APPROVED** [review]
> Reviewed 1 of 17 files at r1, 19 of 19 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51240)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @RichardJCai, and @rohany)

___
*[pkg/sql/drop_type.go, line 69 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51240#-MDGXfiv0zhTUspp3L_t:-MDGXfiv0zhTUspp3L_u:byfk09u) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bc4234cfe21854e973f72fdfcc3d21eb3b943ab/pkg/sql/drop_type.go#L69)):*
> ```Go
> 
> 		// Get the array type that needs to be dropped as well.
> 		// TODO (rohany): This should be a method on the desc.Collection.
> ```

What's currently preventing this (here and elsewhere)?
___
*[pkg/sql/catalog/lease/lease.go, line 414 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51240#-MDGavfn1gsDxU_0DHKF:-MDGavfn1gsDxU_0DHKG:b-j34iuc) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bc4234cfe21854e973f72fdfcc3d21eb3b943ab/pkg/sql/catalog/lease/lease.go#L414)):*
> ```Go
> 				}
> 				if descsToUpdate[id] == nil {
> 					return sqlbase.ErrDescriptorNotFound
> ```

Maybe add a comment with the `PublishMultiple` issue here?

**Review by rohany — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51240)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @lucy-zhang, @RichardJCai, and @rohany)

___
*[pkg/sql/drop_type.go, line 69 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51240#-MDGXfiv0zhTUspp3L_t:-MDGwQaaApyuORE8g4i2:bb32uud) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bc4234cfe21854e973f72fdfcc3d21eb3b943ab/pkg/sql/drop_type.go#L69)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

What's currently preventing this (here and elsewhere)?
</blockquote></details>

Nothing, this PR was just out of date. Updated.
___
*[pkg/sql/catalog/lease/lease.go, line 414 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/51240#-MDGavfn1gsDxU_0DHKF:-MDGwUINElOhvx_zUgQx:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bc4234cfe21854e973f72fdfcc3d21eb3b943ab/pkg/sql/catalog/lease/lease.go#L414)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Maybe add a comment with the `PublishMultiple` issue here?
</blockquote></details>

Done.

**Comment by rohany** [comment]
I'll let @ajwerner take a final pass too

**Review by thoszhang — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51240)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @lucy-zhang, @RichardJCai, and @rohany)

___
*[pkg/sql/sqlbase/structured.go, line 3916 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/51240#-MDGybJ-669E7eHhfUlC:-MDGybJ-669E7eHhfUlD:b-oz8o77) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1dbabfaacbe38e54ed7257d5fb03e9849018429/pkg/sql/sqlbase/structured.go#L3916)):*
> ```Go
> 		return t.Table.Dropped()
> 	case *Descriptor_Type:
> 		return t.Type.Dropped()
> ```

Good catch

**Review by ajwerner — APPROVED** [review]
:lgtm:

We'll need to be careful in `DROP DATABASE` to do the right thing when dropping referenced types. 

> Reviewed 2 of 11 files at r4, 1 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51240#-:-MDKoi7CEodcBJ9MznmS:bra867j)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @lucy-zhang, @RichardJCai, and @rohany)

**Comment by rohany** [comment]
> We'll need to be careful in DROP DATABASE to do the right thing when dropping referenced types.

see https://github.com/cockroachdb/cockroach/pull/51362 :)

**Pull request merged**